core/src/apps/bitcoin/get_public_key.py

```diff
@@ -2,21 +2,39 @@
 
 from trezor import wire
 from trezor.enums import InputScriptType
-from trezor.messages import HDNodeType, PublicKey
+from trezor.messages import AuthorizeCoinJoin, HDNodeType, PublicKey, UnlockPath
 
 from apps.common import coininfo, paths
-from apps.common.keychain import get_keychain
+from apps.common.keychain import FORBIDDEN_KEY_PATH, get_keychain
+
+from . import authorization
 
 if TYPE_CHECKING:
     from trezor.messages import GetPublicKey
+    from trezor.protobuf import MessageType
 
 
-async def get_public_key(ctx: wire.Context, msg: GetPublicKey) -> PublicKey:
+async def get_public_key(
+    ctx: wire.Context, msg: GetPublicKey, auth_msg: MessageType | None = None
+) -> PublicKey:
     coin_name = msg.coin_name or "Bitcoin"
     script_type = msg.script_type or InputScriptType.SPENDADDRESS
     coin = coininfo.by_name(coin_name)
     curve_name = msg.ecdsa_curve_name or coin.curve_name
 
+    # Require preauthorization to access SLIP25 paths.
+    if msg.address_n and msg.address_n[0] == paths.SLIP25_PURPOSE:
+        if auth_msg is not None and AuthorizeCoinJoin.is_type_of(auth_msg):
+            if not authorization.from_cached_message(auth_msg).check_get_public_key(
+                msg
+            ):
+                raise wire.ProcessError("Unauthorized operation")
+        elif auth_msg is not None and UnlockPath.is_type_of(auth_msg):
+            if auth_msg.address_n != msg.address_n[: len(auth_msg.address_n)]:
+                raise FORBIDDEN_KEY_PATH
+        else:
+            raise FORBIDDEN_KEY_PATH
```

Comment:
this would be much nicer linearized in a standalone function:
```
def should_unlock_slip25_path(address_n, auth_msg):
     if not address_n or address_n[0] != SLIP25:
          return False
     if auth_msg is None:
          return False
     ...
```

---

Optionally, you can make a custom schema object:
```
class Slip25AwareSchema:
     @staticmethod
     def match(path: Bip32Path) -> bool:
            return not path or path[0] != paths.SLIP25_PURPOSE

if authorization.is_slip25_unlocked(auth_msg):
      schema = paths.AlwaysMatchingSchema
else:
      schema = Slip25AwareSchema
```
or even integrate the auth msg and the unlock check into it. But the "if not unlocked: raise FORBIDDEN_KEY_PATH" is working here nicely, toying with schemas would only start being useful if used in more places.

Reply:
TBH I don't understand the proposed simplification. Since I removed the option to call GetPublicKey with DoPreauthorized in https://github.com/trezor/trezor-firmware/commit/e0803c17fa2cdb54d1bb17505fb3dfab48f2e127 it seems simple enough now.